### PR TITLE
Add Distributed test and supported multi device unittest

### PR DESCRIPTION
**Pull request by karan6181**
## Description
- Added Distributed test class to simulate the distributed workflow on a single machine
- Added multi-device dataloader unittest
- Run pytest sequentially due to mutli-processing issue with `-n auto`

### CODE REVIEW

File: tests/common/distributed.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        # Fetch custom pytest marker `world_size`
+        # and override with a class parameter
+        for mark in getattr(request.function, 'pytestmark', []):
+            if mark.name == 'world_size':
+                self.world_size = mark.args[0]
+
+    def get_free_tcp_port(self) -> int:
+        """Get a free socket port to listen on."""
+        # from https://www.programcreek.com/python/?CodeExample=get+free+port
+        tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (knighton) [inline tests/common/distributed.py:103]: perhaps just socket.socket() as those args are the defaults? url mention not necessary :)

File: tests/test_distributed.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        assert dist.get_rank() < 4
+
+
+# @pytest.mark.skip
+class TestInit(DistributedTest):
+    world_size = 2
+
+    @pytest.mark.parametrize('batch_size', [128])
+    @pytest.mark.parametrize('drop_last', [False, True])
+    @pytest.mark.parametrize('num_workers', [1, 8])
```
Comment (knighton) [inline tests/test_distributed.py:44]: Should we add zero here?

...('num_workers', [0, 1, 8])?

Comment (karan6181) [reply tests/test_distributed.py]: Added in https://github.com/mosaicml/streaming/pull/57/commits/8741970777041113ee42870ed439f33131211835.

File: tests/common/distributed.py (hunk 1 above)
Comment (karan6181) [reply tests/common/distributed.py]: Changed in https://github.com/mosaicml/streaming/pull/57/commits/8741970777041113ee42870ed439f33131211835.

**Pull request merged**